### PR TITLE
Gradient clipping + automatic max_norm, fix double scale.update

### PR DESCRIPTION
**Pull request by terrainer**
No underflow issues detected from the init_scale value as of yet. trained for 6 hours with no visible issues.

### CODE REVIEW

**Comment by terrainer** [comment]
### More details:

**GradScaler:**
This also adds all the relevant gradscaler tweaks (default values, besides `init_scale`). I was thinking it could be possible to allow users to tweak this in their options file (as well as a gradient clipping toggle). The `init_scale` doesn't necessarily need to change from your value `2**11`, as long as there are no stability issues - I'm sure you're already aware that gradscaler/autocast adjusts itself for infs and NaNs, but the important thing is to make sure that the inital scaling doesn't blow out training entirely.

**Double _scaler.update()_ call:**
From [PyTorch documentation](https://pytorch.org/docs/stable/notes/amp_examples.html#working-with-multiple-models-losses-and-optimizers):

> If your network has multiple losses, you must call [scaler.scale](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.scale) on each of them individually. If your network has multiple optimizers, you may call [scaler.unscale_](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.unscale_) on any of them individually, and you must call [scaler.step](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.step) on each of them individually.
> 
> However, [scaler.update](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.update) should only be called once, after all optimizers used this iteration have been stepped

**Gradient Clipping:**
In theory, this unscales as required, uses a heuristic approach to automatically choose a max_norm value, then clips the gradient with that value. I'm not sure if this is working as intended, as there's currently no way to monitor the gradients and weights implemented (to my knowledge).

**Comment by muslll** [comment]
Removed one of the scaler.update() in [this commit](https://github.com/muslll/neosr/commit/39f9a4a28c538c600dcef1a4725c8ecc95357145). Other changes were not tested yet. Can you modify the PR to contain only the gradient clipping part? I will have to verify if the implementation is correct to the pytorch documentation, but in case it is and testing proves it works, I can merge the PR directly.

**Comment by terrainer** [comment]
Done!

**Comment by terrainer** [comment]
Updated with the a new & better adaptive grad clipping implementation
